File: CBLClient/Apps/JavaRequestHandlers/src/main/java/com/couchbase/mobiletestkit/javacommon/RequestHandler/PeerToPeerRequestHandler.java
package com.couchbase.mobiletestkit.javacommon.RequestHandler;

/*
  Created by sridevi.saragadam on 7/9/18.
 */

import java.io.IOException;
import java.net.URI;
import java.security.cert.Certificate;
import java.security.cert.X509Certificate;
import java.util.ArrayList;
import java.util.List;
import com.couchbase.lite.ClientCertificateAuthenticator;
import com.couchbase.lite.ConflictResolver;
import com.couchbase.lite.CouchbaseLiteException;
import com.couchbase.lite.ListenerAuthenticator;
import com.couchbase.lite.ListenerCertificateAuthenticator;
import com.couchbase.lite.TLSIdentity;
import com.couchbase.lite.URLEndpointListener;
import com.couchbase.lite.URLEndpointListenerConfiguration;
import com.couchbase.lite.*;
import com.couchbase.mobiletestkit.javacommon.Args;
import com.couchbase.mobiletestkit.javacommon.RequestHandlerDispatcher;
import com.couchbase.mobiletestkit.javacommon.util.Log;


public class PeerToPeerRequestHandler implements MessageEndpointDelegate {
    private static final String TAG = "P2PHANDLER";

    final ReplicatorRequestHandler replicatorRequestHandlerObj = new ReplicatorRequestHandler();

    public void clientStart(Args args) {
        Replicator replicator = args.get("replicator");
        replicator.start();
        Log.i(TAG, "Replication started .... ");
    }

    public Replicator configure(Args args) throws Exception {
        String ipaddress = args.get("host");
        int port = args.get("port");
        Database sourceDb = args.get("database");
        String serverDBName = args.get("serverDBName");
        String replicationType = args.get("replicationType");
        Boolean continuous = args.get("continuous");
        String endPointType = args.get("endPointType");
        List<String> documentIds = args.get("documentIDs");
        Boolean push_filter = args.get("push_filter");
        Boolean pull_filter = args.get("pull_filter");
        String filter_callback_func = args.get("filter_callback_func");
        String conflict_resolver = args.get("conflict_resolver");
        Boolean disableTls = args.get("tls_disable");
        String tlsAuthType = args.get("tls_auth_type");
        Boolean serverVerificationMode = args.get("server_verification_mode");
        Boolean tlsAuthenticator = args.get("tls_authenticator");
        String heartbeat = args.get("heartbeat");
        String maxRetries = args.get("max_retries");
        String maxTimeout = args.get("max_timeout");

        ReplicatorConfiguration config;
        Replicator replicator;
        URI uri;

        if (replicationType == null) {
            replicationType = "push_pull";
        }
        replicationType = replicationType.toLowerCase();
        ReplicatorConfiguration.ReplicatorType replType;
        if (replicationType.equals("push")) {
            replType = ReplicatorConfiguration.ReplicatorType.PUSH;
        } else if (replicationType.equals("pull")) {
            replType = ReplicatorConfiguration.ReplicatorType.PULL;
        } else {
            replType = ReplicatorConfiguration.ReplicatorType.PUSH_AND_PULL;
        }
        Log.i(TAG, "serverDBName is " + serverDBName);
        if (disableTls) {
            uri = new URI("ws://" + ipaddress + ":" + port + "/" + serverDBName);
        } else {
            uri = new URI("wss://" + ipaddress + ":" + port + "/" + serverDBName);
        }

        if (endPointType.equals("URLEndPoint")) {
            URLEndpoint urlEndPoint = new URLEndpoint(uri);
            config = new ReplicatorConfiguration(sourceDb, urlEndPoint);
        } else if (endPointType.equals("MessageEndPoint")) {
            MessageEndpoint messageEndPoint = new MessageEndpoint("p2p", uri, ProtocolType.BYTE_STREAM, this);
            config = new ReplicatorConfiguration(sourceDb, messageEndPoint);
        } else {
            throw new IllegalArgumentException("Incorrect EndPoint type");
        }
        config.setReplicatorType(replType);
        if (continuous != null) {
            config.setContinuous(continuous);
        } else {
            config.setContinuous(false);
        }
        if (documentIds != null) {
            config.setDocumentIDs(documentIds);
        }
        if (heartbeat != null && !heartbeat.trim().isEmpty()){
            config.setHeartbeat(Integer.parseInt(heartbeat));
        }

        if (maxRetries != null && !maxRetries.trim().isEmpty()) {
            config.setMaxAttempts(Integer.parseInt(maxRetries));
        }

        if (maxTimeout != null && !maxTimeout.trim().isEmpty()) {
            config.setMaxAttemptWaitTime(Integer.parseInt(maxTimeout));
        }


        if (push_filter) {
            switch (filter_callback_func) {
                case "boolean":
                    config.setPushFilter(new ReplicatorBooleanFilterCallback());
                    break;
                case "deleted":
                    config.setPushFilter(new ReplicatorDeletedFilterCallback());
                    break;
                case "access_revoked":
                    config.setPushFilter(new ReplicatorAccessRevokedFilterCallback());
                    break;
                default:
                    config.setPushFilter(new DefaultReplicatorFilterCallback());
                    break;
            }
        }
        if (pull_filter) {
            switch (filter_callback_func) {
                case "boolean":
                    config.setPullFilter(new ReplicatorBooleanFilterCallback());
                    break;
                case "deleted":
                    config.setPullFilter(new ReplicatorDeletedFilterCallback());
                    break;
                case "access_revoked":
                    config.setPullFilter(new ReplicatorAccessRevokedFilterCallback());
                    break;
                default:
                    config.setPullFilter(new DefaultReplicatorFilterCallback());
                    break;
            }
        }
        if (args.get("basic_auth") != null) {
            config.setAuthenticator(args.get("basic_auth"));
        }

        if (tlsAuthType.equals("self_signed")) {
            TLSIdentity tlsIdentity = RequestHandlerDispatcher.context.getSelfSignedIdentity();
            if (tlsIdentity != null) {
                List<Certificate> certs = tlsIdentity.getCerts();
                X509Certificate cert = (X509Certificate) certs.get(0);
                config.setPinnedServerCertificate(cert.getEncoded());
                Log.i(TAG, "Pinned the certs ... .... ");
            }
        }

        if (tlsAuthenticator) {
            TLSIdentity identity = RequestHandlerDispatcher.context.getClientCertsIdentity();
            ClientCertificateAuthenticator clientCertificateAuthenticator = new ClientCertificateAuthenticator(identity);
            config.setAuthenticator(clientCertificateAuthenticator);
        }

        if (serverVerificationMode) {
            config.setAcceptOnlySelfSignedServerCertificate(true);
        }
        switch (conflict_resolver) {
            case "local_wins":
                config.setConflictResolver(new LocalWinsCustomConflictResolver());
                break;
            case "remote_wins":
                config.setConflictResolver(new RemoteWinsCustomConflictResolver());
                break;
            case "null":
                config.setConflictResolver(new NullCustomConflictResolver());
                break;
            case "merge":
                config.setConflictResolver(new MergeCustomConflictResolver());
                break;
            case "incorrect_doc_id":
                config.setConflictResolver(new IncorrectDocIdConflictResolver());
                break;
            case "delayed_local_win":
                config.setConflictResolver(new DelayedLocalWinConflictResolver());
                break;
            case "delete_not_win":
                config.setConflictResolver(new DeleteDocConflictResolver());
                break;
            case "exception_thrown":
                config.setConflictResolver(new ExceptionThrownConflictResolver());
                break;
            default:
                config.setConflictResolver(ConflictResolver.DEFAULT);
                break;
        }
        replicator = new Replicator(config);
        return replicator;
    }

    class MyReplicatorListener implements ReplicatorChangeListener {
        private final List<ReplicatorChange> changes = new ArrayList<>();

        public List<ReplicatorChange> getChanges() {
            return changes;
        }

        @Override
        public void changed(ReplicatorChange change) {
            changes.add(change);
        }
    }

    public URLEndpointListener serverStart(Args args) throws IOException, CouchbaseLiteException {
        int port = args.get("port");
        Database sourceDb = args.get("database");
        URLEndpointListenerConfiguration config = new URLEndpointListenerConfiguration(sourceDb);
        Boolean disableTls = args.get("tls_disable");
        Boolean tlsAuthenticator = args.get("tls_authenticator");
        String tlsAuthType = args.get("tls_auth_type");
        System.out.println(tlsAuthType);

        if (port > 0) {
            port = args.get("port");
            config.setPort(port);
        }
        config.setDisableTls(disableTls);

        if (args.get("basic_auth") != null) {
            ListenerAuthenticator listenerAuthenticator = args.get("basic_auth");
            config.setAuthenticator(listenerAuthenticator);
        }

        if (tlsAuthType.equals("self_signed_create")) {
            TLSIdentity identity = RequestHandlerDispatcher.context.getCreateIdentity();
            config.setTlsIdentity(identity);
        }

        if (tlsAuthType.equals("self_signed")) {
            TLSIdentity identity = RequestHandlerDispatcher.context.getSelfSignedIdentity();
            config.setTlsIdentity(identity);
            Log.e(TAG,"ServerSide setting the identity");
        }
        if (tlsAuthenticator) {
            List<Certificate> certsList = RequestHandlerDispatcher.context.getAuthenticatorCertsList();
            ListenerCertificateAuthenticator listenerCertificateAuthenticator = new ListenerCertificateAuthenticator(certsList);
            config.setAuthenticator(listenerCertificateAuthenticator);
        }

        URLEndpointListener p2ptcpListener = new URLEndpointListener(config);
        p2ptcpListener.start();
        System.out.println(p2ptcpListener.getPort());
        return p2ptcpListener;
    }

    public int getListenerPort(Args args) {
        URLEndpointListener p2ptcpListener = args.get("listener");
        return p2ptcpListener.getPort();
    }

    public ReplicatorTcpListener messageEndpointListenerStart(Args args) throws IOException {
        Database sourceDb = args.get("database");
        int port = args.get("port");
        MessageEndpointListener messageEndpointListener =
                new MessageEndpointListener(new MessageEndpointListenerConfiguration(
                        sourceDb,
                        ProtocolType.BYTE_STREAM));
        ReplicatorTcpListener p2ptcpListener = new ReplicatorTcpListener(sourceDb, port);
        p2ptcpListener.start();
        return p2ptcpListener;
    }

    public void serverStop(Args args) {
        String endPointType = args.get("endPointType");
        if (endPointType.equals("MessageEndPoint")) {
            ReplicatorTcpListener p2ptcpListener = args.get("listener");
            p2ptcpListener.stop();
        } else {
            URLEndpointListener p2ptcpListener = args.get("listener");
            p2ptcpListener.stop();
        }
    }

    public MessageEndpointConnection createConnection(MessageEndpoint endpoint) {
        URI url = (URI) endpoint.getTarget();
        return new ReplicatorTcpClientConnection(url);
    }

    public MyDocumentReplicatorListener addReplicatorEventChangeListener(Args args) {
        return replicatorRequestHandlerObj.addReplicatorEventChangeListener(args);
    }

    public void removeReplicatorEventListener(Args args) {
        replicatorRequestHandlerObj.removeReplicatorEventListener(args);
    }

    public int changeListenerChangesCount(Args args) {
        return replicatorRequestHandlerObj.changeListenerChangesCount(args);
    }

    public List<String> replicatorEventGetChanges(Args args) {
        return replicatorRequestHandlerObj.replicatorEventGetChanges(args);
    }
}

File: CBLClient/Apps/CBLTestServer-iOS/CBLTestServer-iOS/Server/PeerToPeerRequestHandler.swift
//
//  PeerToPeerRequestHandler.swift
//  CBLTestServer-iOS
//
//  Created by Sridevi Saragadam on 7/23/18.
//  Copyright © 2018 Raghu Sarangapani. All rights reserved.
//
import Foundation
import CouchbaseLiteSwift

public class PeerToPeerRequestHandler {
    public static let VOID: String? = nil
    fileprivate var _pushPullReplListener:NSObjectProtocol?
    private func _replicatorBooleanFilterCallback(document: Document, flags: DocumentFlags) -> Bool {
        let key:String = "new_field_1"
        if document.contains(key){
            let value:Bool = document.boolean(forKey: key)
            return value;
        }
        return true
    }

    private func _defaultReplicatorFilterCallback(document: Document, flags: DocumentFlags) -> Bool {
        return true;
    }

    private func _replicatorDeletedFilterCallback(document: Document, documentFlags: DocumentFlags) -> Bool {
        if (documentFlags.rawValue == 1) {
            return false
        }
        return true
    }

    private func _replicatorAccessRevokedCallback(document: Document, documentFlags: DocumentFlags) -> Bool {
        if (documentFlags.rawValue == 2) {
            return false
        }
        return true
    }
    
    func dataFromResource(name: String, ofType type: String) throws -> Data {
        let path = Bundle(for: Swift.type(of:self)).path(forResource: name, ofType: type)
        return try! NSData(contentsOfFile: path!, options: []) as Data
    }

    public func handleRequest(method: String, args: Args) throws -> Any? {
        
        switch method {
            
            /////////////////////////////
            // Peer to Peer Apis //
            ////////////////////////////
        #if COUCHBASE_ENTERPRISE
        case "peerToPeer_messageEndpointListenerStart":
            let database: Database = args.get(name:"database")!
            let port: Int = args.get(name:"port")!
            let peerToPeerListener: ReplicatorTcpListener = ReplicatorTcpListener(databases: [database], port: UInt32(port))
            peerToPeerListener.start()
            return peerToPeerListener

        case "peerToPeer_getListenerPort":
            let listener: URLEndpointListener = args.get(name:"listener")!
            return listener.port
    
        case "peerToPeer_serverStart":
            var peerToPeerListener: URLEndpointListener?
            let database: Database = args.get(name:"database")!
            let wsPort: Int = args.get(name: "port")!
            let tls_disable: Bool? = args.get(name: "tls_disable")!
            var config = URLEndpointListenerConfiguration.init(database: [database][0])
            let tlsAuthType: String = args.get(name:"tls_auth_type")!
            let tlsAuthenticator: Bool? = args.get(name: "tls_authenticator")!
            let enable_delta_sync: Bool? = args.get(name: "enable_delta_sync")!

            config.port = UInt16(wsPort)
            config.disableTLS = tls_disable!
            
            if (tlsAuthenticator != false) {
                let rootCertData = try dataFromResource(name: "identity/client-ca", ofType: "der")
                let rootCert = SecCertificateCreateWithData(kCFAllocatorDefault, rootCertData as CFData)!
                let listenerAuth = ListenerCertificateAuthenticator.init(rootCerts: [rootCert])
                try! TLSIdentity.deleteIdentity(withLabel: "CBL-Cert")
                config.authenticator = listenerAuth
                print("========== Setting Authenticator ===========")
            }
            
            if tlsAuthType == "self_signed" {
                let serverCertData = try! dataFromResource(name: "identity/certs", ofType: "p12")
                try! TLSIdentity.deleteIdentity(withLabel: "CBL-Cert")
                let identity = try! TLSIdentity.importIdentity(withData: serverCertData, password: "123", label: "CBL-Cert")
                config.tlsIdentity = identity
                print("============= Setting identity ================")

            } else if tlsAuthType == "self_signed_create" {
                try! TLSIdentity.deleteIdentity(withLabel: "CBL-Cert")
                let id = try! TLSIdentity.createIdentity(forServer: true , attributes: [certAttrCommonName: "CBL-Server"], expiration: nil, label: "CBL-Cert")
                config.tlsIdentity = id
                print("========== Setting CreateIdentity =========")
            }

            if let auth: ListenerPasswordAuthenticator = args.get(name: "basic_auth") {
               config.authenticator = auth
            }

            config.enableDeltaSync = enable_delta_sync!
            peerToPeerListener = URLEndpointListener.init(config: config)
            try peerToPeerListener?.start()
            print("Server is getting started")
            return peerToPeerListener
            
        case "peerToPeer_serverStop":
            let type: String? = args.get(name:"endPointType")!
            if (type == "MessageEndPoint") {
                let listener_obj: ReplicatorTcpListener = args.get(name: "listener")!
                listener_obj.stop()
            } else {
                let listener_obj: URLEndpointListener = args.get(name: "listener")!
                listener_obj.stop()
            }
            
        case "peerToPeer_configure":
            let host: String = args.get(name:"host")!
            let port: Int = args.get(name:"port")!
            let serverDBName: String = args.get(name:"serverDBName")!
            let database: Database = args.get(name:"database")!
            let continuous: Bool? = args.get(name:"continuous")!
            let replication_type: String? = args.get(name: "replicationType")!
            let documentIDs: [String]? = args.get(name: "documentIDs")
            let endPointType: String = args.get(name: "endPointType")!
            let pull_filter: Bool? = args.get(name: "pull_filter")!
            let push_filter: Bool? = args.get(name: "push_filter")!
            let conflict_resolver: String? = args.get(name: "conflict_resolver")!
            let filter_callback_func: String? = args.get(name: "filter_callback_func")
            let auth: Authenticator? = args.get(name: "basic_auth")!
            let heartbeat: String? = args.get(name: "heartbeat")
            
            let serverVerificationMode: Bool? = args.get(name: "server_verification_mode")!
            let tls_disable: Bool? = args.get(name: "tls_disable")!
            let tlsAuthType: String? = args.get(name: "tls_auth_type")!
            let tlsAuthenticator: Bool? = args.get(name: "tls_authenticator")!
            var replicatorConfig: ReplicatorConfiguration
            var replicatorType = ReplicatorType.pushAndPull
            var wsPort: String?
            let maxRetries: String? = args.get(name: "max_retries")
            let maxRetryWaitTime: String? = args.get(name: "max_timeout")

            if let type = replication_type {
                if type == "push" {
                    replicatorType = .push
                } else if type == "pull" {
                    replicatorType = .pull
                } else {
                    replicatorType = .pushAndPull
                }
            }
            
            if tls_disable != true {
               wsPort = "wss"
            } else {
                wsPort = "ws"
            }

            let url = URL(string: "\(wsPort ?? "ws")://\(host):\(port)/\(serverDBName)")!
            if endPointType == "URLEndPoint"{
                let urlEndPoint: URLEndpoint = URLEndpoint(url: url)
                replicatorConfig = ReplicatorConfiguration(database: database, target: urlEndPoint)
            }
            else{
                let endpoint = MessageEndpoint(uid: url.absoluteString, target: url, protocolType: ProtocolType.byteStream, delegate: self)
                replicatorConfig = ReplicatorConfiguration(database: database, target: endpoint)
            }
            
            if auth != nil {
               replicatorConfig.authenticator = auth
            }
            if serverVerificationMode != false {
                replicatorConfig.acceptOnlySelfSignedServerCertificate = true
            }
            
            if tlsAuthType == "self_signed" {
                let certData = try! dataFromResource(name: "identity/certs", ofType: "p12")
                try! TLSIdentity.deleteIdentity(withLabel: "CBL-Cert")
                let identity = try! TLSIdentity.importIdentity(withData: certData, password: "123", label: "CBL-Cert")
                replicatorConfig.pinnedServerCertificate = identity.certs[0]
                print("======= pinned the certs to Replicator ============")
            }
            if tlsAuthenticator != false {
                let clientCertData = try dataFromResource(name: "identity/client", ofType: "p12")
                try! TLSIdentity.deleteIdentity(withLabel: "CBL-Cert")
                let identity = try TLSIdentity.importIdentity(withData: clientCertData, password: "123", label: "CBL-Cert")
                replicatorConfig.authenticator = ClientCertificateAuthenticator(identity: identity)
                print("====== Added Autheticator to Replicator ========")
            }

            if continuous != nil {
                replicatorConfig.continuous = continuous!
            }
            if documentIDs != nil {
                replicatorConfig.documentIDs = documentIDs
            }
            if pull_filter != false {
                if filter_callback_func == "boolean" {
                    replicatorConfig.pullFilter = _replicatorBooleanFilterCallback;
                } else if filter_callback_func == "deleted" {
                    replicatorConfig.pullFilter = _replicatorDeletedFilterCallback;
                } else if filter_callback_func == "access_revoked" {
                    replicatorConfig.pullFilter = _replicatorAccessRevokedCallback;
                } else {
                    replicatorConfig.pullFilter = _defaultReplicatorFilterCallback;
                }
            }
            if push_filter != false {
                if filter_callback_func == "boolean" {
                    replicatorConfig.pushFilter = _replicatorBooleanFilterCallback;
                } else if filter_callback_func == "deleted" {
                    replicatorConfig.pushFilter = _replicatorDeletedFilterCallback;
                } else if filter_callback_func == "access_revoked" {
                    replicatorConfig.pushFilter = _replicatorAccessRevokedCallback;
                } else {
                    replicatorConfig.pushFilter = _defaultReplicatorFilterCallback;
                }
            }
            switch conflict_resolver {
                case "local_wins":
                    replicatorConfig.conflictResolver = LocalWinCustomConflictResolver();
                    break
                case "remote_wins":
                    replicatorConfig.conflictResolver = RemoteWinCustomConflictResolver();
                    break;
                case "null":
                    replicatorConfig.conflictResolver = NullWinCustomConflictResolver();
                    break;
                case "merge":
                    replicatorConfig.conflictResolver = MergeWinCustomConflictResolver();
                    break;
                case "incorrect_doc_id":
                    replicatorConfig.conflictResolver = IncorrectDocIdCustomConflictResolver();
                    break;
                case "delayed_local_win":
                    replicatorConfig.conflictResolver = DelayedLocalWinCustomConflictResolver();
                    break;
                case "exception_thrown":
                    replicatorConfig.conflictResolver = ExceptionThrownCustomConflictResolver();
                    break;
                default:
                    replicatorConfig.conflictResolver = ConflictResolver.default
                    break;
            }
            if let heartbeat = heartbeat, let heartbeatDouble = Double(heartbeat) {
                replicatorConfig.heartbeat = heartbeatDouble
            }
            if let maxRetries = maxRetries, let maxRetryInInt = UInt(maxRetries) {
                replicatorConfig.maxAttempts = maxRetryInInt
            }
            if let maxRetryWaitTime = maxRetryWaitTime, let maxRetryWaitTimeDouble = Double(maxRetryWaitTime) {
                replicatorConfig.maxAttemptWaitTime = maxRetryWaitTimeDouble
            }
            replicatorConfig.replicatorType = replicatorType
            let replicator: Replicator = Replicator(config: replicatorConfig)
            return replicator

        case "peerToPeer_clientStart":
            let replicator: Replicator = args.get(name:"replicator")!
            replicator.start()
            print("Replicator has started")
            
        case "peerToPeer_addReplicatorEventChangeListener":
            let replication_obj: Replicator = args.get(name: "replicator")!
            let changeListener = MyDocumentReplicationListener()
            let listenerToken = replication_obj.addDocumentReplicationListener(changeListener.listener)
            changeListener.listenerToken = listenerToken
            return changeListener

        case "peerToPeer_removeReplicatorEventListener":
            let replication_obj: Replicator = args.get(name: "replicator")!
            let changeListener : MyDocumentReplicationListener = (args.get(name: "changeListener"))!
            replication_obj.removeChangeListener(withToken: changeListener.listenerToken!)

        case "peerToPeer_replicatorEventChangesCount":
            let changeListener: MyDocumentReplicationListener = (args.get(name: "changeListener"))!
            return changeListener.getChanges().count

        case "peerToPeer_replicatorEventGetChanges":
            let changeListener: MyDocumentReplicationListener = (args.get(name: "changeListener"))!
            let changes: [DocumentReplication] = changeListener.getChanges()
            var event_list: [String] = []
            for change in changes {
                for document in change.documents {
                    let doc_event:String = "doc_id: " + document.id
                    let error:String = ", error_code: " + document.error.debugDescription + ", error_domain: nil"
                    let flags:String = ", flags: " + document.flags.rawValue.description
                    let push:String = ", push: " + change.isPush.description
                    event_list.append(doc_event + error + push + flags)
                }
            }
            return event_list
        #endif
        default:
            throw RequestHandlerError.MethodNotFound(method)
        }
        return PeerToPeerRequestHandler.VOID
    }
}
#if COUCHBASE_ENTERPRISE
extension PeerToPeerRequestHandler: MessageEndpointDelegate {
    public func createConnection(endpoint: MessageEndpoint) -> MessageEndpointConnection {
        let url = endpoint.target as! URL
        return ReplicatorTcpClientConnection.init(url: url)
    }
}
#endif

private class LocalWinCustomConflictResolver: ConflictResolverProtocol {
    func resolve(conflict: Conflict) -> Document? {
        let localDoc = conflict.localDocument!
        let remoteDoc = conflict.remoteDocument!
        let docId = conflict.documentID
        checkMismatchDocId(localDoc: localDoc, remoteDoc: remoteDoc, docId: docId)
        return localDoc
    }
}

private class RemoteWinCustomConflictResolver: ConflictResolverProtocol{
    func resolve(conflict: Conflict) -> Document? {
        let localDoc = conflict.localDocument!
        let remoteDoc = conflict.remoteDocument!
        let docId = conflict.documentID
        checkMismatchDocId(localDoc: localDoc, remoteDoc: remoteDoc, docId: docId)
        return remoteDoc
    }
}

private class NullWinCustomConflictResolver: ConflictResolverProtocol{
    func resolve(conflict: Conflict) -> Document? {
        let localDoc = conflict.localDocument!
        let remoteDoc = conflict.remoteDocument!
        let docId = conflict.documentID
        checkMismatchDocId(localDoc: localDoc, remoteDoc: remoteDoc, docId: docId)
        return nil
    }
}

private class MergeWinCustomConflictResolver: ConflictResolverProtocol{
    func resolve(conflict: Conflict) -> Document? {
        let localDoc = conflict.localDocument!
        let remoteDoc = conflict.remoteDocument!
        let docId = conflict.documentID
        checkMismatchDocId(localDoc: localDoc, remoteDoc: remoteDoc, docId: docId)
        let newDoc = localDoc.toMutable()
        let remoteDocMap = remoteDoc.toDictionary()
        for (key, value) in remoteDocMap {
            if !newDoc.contains(key) {
                newDoc.setValue(value, forKey: key)
            }
        }
        return newDoc
    }
}

private class DelayedLocalWinCustomConflictResolver: ConflictResolverProtocol{
    func resolve(conflict: Conflict) -> Document? {
        let localDoc = conflict.localDocument!
        let remoteDoc = conflict.remoteDocument!
        let docId = conflict.documentID
        checkMismatchDocId(localDoc: localDoc, remoteDoc: remoteDoc, docId: docId)
        sleep(10)
        return localDoc
    }
}

private class IncorrectDocIdCustomConflictResolver: ConflictResolverProtocol{
    func resolve(conflict: Conflict) -> Document? {
        let localDoc = conflict.localDocument!
        let remoteDoc = conflict.remoteDocument!
        let docId = conflict.documentID
        checkMismatchDocId(localDoc: localDoc, remoteDoc: remoteDoc, docId: docId)
        let newId = "changed\(docId)"
        let newDoc = MutableDocument(id: newId, data: localDoc.toDictionary())
        newDoc.setValue("_id", forKey: newId)
        return newDoc
    }
}

private class DeleteDocCustomConflictResolver: ConflictResolverProtocol{
    func resolve(conflict: Conflict) -> Document? {
        let localDoc = conflict.localDocument
        let remoteDoc = conflict.remoteDocument
        let docId = conflict.documentID
        checkMismatchDocId(localDoc: localDoc, remoteDoc: remoteDoc, docId: docId)
        if remoteDoc == nil {
            return localDoc
        }
        return nil
    }
}

private class ExceptionThrownCustomConflictResolver: ConflictResolverProtocol{
    func resolve(conflict: Conflict) -> Document? {
        let localDoc = conflict.localDocument
        let remoteDoc = conflict.remoteDocument
        let docId = conflict.documentID
        checkMismatchDocId(localDoc: localDoc, remoteDoc: remoteDoc, docId: docId)
        NSException(name: .internalInconsistencyException,
                    reason: "some exception happened inside custom conflict resolution",
                    userInfo: nil).raise()
        return localDoc
    }
}

private func checkMismatchDocId(localDoc: Document?, remoteDoc: Document?, docId: String) -> Void{
    if let remoteDocId = remoteDoc?.id {
        if remoteDocId != docId {
            NSException(name: .internalInconsistencyException,
                        reason: "DocId mismatch",
                        userInfo: nil).raise()
        }
    }
    if let localDocId = localDoc?.id {
        if localDocId != docId {
            NSException(name: .internalInconsistencyException,
                        reason: "DocId mismatch",
                        userInfo: nil).raise()
        }
    }
}

